File: compressors.py
"""
Lossless compressors for source code and terminal output.

"Lossless" here has a precise, testable meaning:

* ``lossless_code_compressor``  -> the compressed source parses to an
  **identical AST** as the input. Only human-facing comments and redundant
  blank space are removed. Indentation, string contents (including
  docstrings) and every executable token are preserved byte-for-byte.
* ``lossless_terminal_cleaner`` -> every line carrying *signal* (errors,
  failures, completion markers) survives. Only mechanical progress noise
  (tickers, percentages, download loops, spinner frames) is dropped, and the
  drop is accounted for so nothing silently disappears.

Both functions are pure: same input -> same output, no globals, no I/O.
"""

from __future__ import annotations

import re
from typing import Iterable, List, Optional, Sequence, Tuple

__all__ = [
    "lossless_code_compressor",
    "lossless_terminal_cleaner",
]


# ---------------------------------------------------------------------------
# Code compression
# ---------------------------------------------------------------------------

# Comments that are NOT human prose: they change how tools/interpreters behave.
# Stripping these would be lossy, so they are always kept.
_SEMANTIC_COMMENT = re.compile(
    r"""^\#\s*(?:
          !                      # shebang
        | -\*-                   # -*- coding: ... -*-
        | (?:type|mypy|pyright)\b
        | noqa\b
        | pragma\b
        | pylint\b
        | flake8\b
        | ruff\b
        | fmt\s*:\s*(?:on|off|skip)
        | yapf\s*:\s*(?:disable|enable)
        | isort\s*:\s*
        | nosec\b
        | coding[:=]
    )""",
    re.VERBOSE | re.IGNORECASE,
)


def _scan_line(line: str, in_triple: Optional[str]) -> Tuple[Optional[int], Optional[str]]:
    """Scan one physical line, tracking string state.

    Returns ``(comment_start, new_triple_state)`` where ``comment_start`` is the
    index of the ``#`` that begins a real comment, or ``None`` if this line has
    no comment. A ``#`` inside a string literal is never reported.
    """
    i = 0
    n = len(line)

    # Continue an open triple-quoted string from a previous line.
    if in_triple is not None:
        while i < n:
            if line[i] == "\\":
                i += 2
                continue
            if line.startswith(in_triple, i):
                i += len(in_triple)
                in_triple = None
                break
            i += 1
        else:
            return None, in_triple  # still inside the string

    while i < n:
        ch = line[i]

        if ch == "#":
            return i, None

        if ch in ("'", '"'):
            triple = line[i : i + 3]
            if triple in ("'''", '"""'):
                i += 3
                # Does it close on this same line?
                while i < n:
                    if line[i] == "\\":
                        i += 2
                        continue
                    if line.startswith(triple, i):
                        i += len(triple)
                        break
                    i += 1
                else:
                    return None, triple  # spills onto the next line
                continue

            # Single-quoted string.
            quote = ch
            i += 1
            while i < n:
                if line[i] == "\\":
                    i += 2
                    continue
                if line[i] == quote:
                    i += 1
                    break
                i += 1
            continue

        i += 1

    return None, None


def lossless_code_compressor(
    raw_code: str,
    *,
    max_consecutive_blank_lines: int = 1,
    strip_trailing_comments: bool = True,
    keep_semantic_comments: bool = True,
) -> str:
    """Strip human comments and collapse blank-line runs, losslessly.

    Parameters
    ----------
    raw_code:
        Source text to compress.
    max_consecutive_blank_lines:
        Runs of blank lines longer than this are collapsed down to it.
        ``0`` removes blank lines between statements entirely.
    strip_trailing_comments:
        Also remove ``code  # comment`` tails, not just whole-line comments.
    keep_semantic_comments:
        Preserve shebangs, coding cookies and tool directives (``# noqa``,
        ``# type:``, ``# fmt: off`` ...), which are not human prose.

    Guarantees
    ----------
    * Never edits anything inside a string literal (docstrings included).
    * Never changes indentation of a kept line.
    * Blank lines inside triple-quoted strings are left untouched.
    """
    if not raw_code:
        return ""

    lines = raw_code.splitlines()
    out: List[str] = []
    in_triple: Optional[str] = None
    blank_run = 0

    for lineno, line in enumerate(lines, start=1):
        was_in_string = in_triple is not None
        comment_start, in_triple = _scan_line(line, in_triple)

        # Lines that live inside a multi-line string are sacred: emit verbatim.
        if was_in_string:
            out.append(line)
            blank_run = 0
            continue

        if comment_start is not None:
            comment_text = line[comment_start:]
            stripped_before = line[:comment_start].strip()
            is_whole_line_comment = stripped_before == ""

            keep_it = keep_semantic_comments and (
                _SEMANTIC_COMMENT.match(comment_text.strip())
                or (lineno == 1 and comment_text.startswith("#!"))
            )

            if keep_it:
                out.append(line.rstrip())
                blank_run = 0
                continue

            if is_whole_line_comment:
                continue  # drop the line entirely

            if strip_trailing_comments:
                line = line[:comment_start].rstrip()
            else:
                line = line.rstrip()
        else:
            line = line.rstrip()

        if line.strip() == "":
            blank_run += 1
            if blank_run > max_consecutive_blank_lines:
                continue
            out.append("")
            continue

        blank_run = 0
        out.append(line)

    # Trim leading/trailing blank lines produced by removals.
    while out and out[0].strip() == "":
        out.pop(0)
    while out and out[-1].strip() == "":
        out.pop()

    return "\n".join(out)


# ---------------------------------------------------------------------------
# Terminal / build-log cleaning
# ---------------------------------------------------------------------------

# Signal always wins: if a line matches these, it is kept no matter what.
_SIGNAL_PATTERNS: Sequence[re.Pattern] = (
    re.compile(r"\b(?:error|errors)\b", re.IGNORECASE),
    re.compile(r"\b(?:fatal|panic|abort(?:ed)?|segfault|core dumped)\b", re.IGNORECASE),
    re.compile(r"\b(?:fail|failed|failure|failing)\b", re.IGNORECASE),
    re.compile(r"\b(?:exception|traceback|stack ?trace|assertion)\b", re.IGNORECASE),
    re.compile(r"\b(?:undefined reference|cannot find|not found|no such file)\b", re.IGNORECASE),
    re.compile(r"\b(?:exit(?:ed)? (?:code|status)|exit code)\b", re.IGNORECASE),
    re.compile(r"\b(?:succeeded|success|successful|completed successfully)\b", re.IGNORECASE),
    re.compile(r"\b(?:build (?:finished|complete|succeeded|failed))\b", re.IGNORECASE),
    re.compile(r"\b(?:finished|done) in\b", re.IGNORECASE),
    re.compile(r"^\s*(?:E|ERROR|FAIL|FAILED|CRITICAL)\b"),
    re.compile(r"\bwarning\b.*\b(?:treated as error|-Werror)\b", re.IGNORECASE),
    re.compile(r"^\s*(?:at |File \"|\s+\^+\s*$)"),  # stack frames / carets
)

# Pure mechanical noise: progress, downloads, spinners.
_NOISE_PATTERNS: Sequence[re.Pattern] = (
    re.compile(r"^\s*\[\s*\d+\s*/\s*\d+\s*\]"),                  # [1/250] Compiling foo.c
    re.compile(r"^\s*\(\s*\d+\s*/\s*\d+\s*\)"),                  # (17/250) Linking
    re.compile(r"^\s*\d{1,3}\s*%"),                              # 45% completed
    re.compile(r"\b\d{1,3}(?:\.\d+)?\s*%\s*(?:complete|completed|done|finished)?\b", re.IGNORECASE),
    re.compile(r"^\s*(?:Compiling|Building|Downloading|Fetching|Extracting|Unpacking|"
               r"Installing|Resolving|Linking|Indexing|Uploading|Pulling|Cloning)\b", re.IGNORECASE),
    re.compile(r"\b(?:ETA|eta)\s+\d"),                           # ETA 00:12
    re.compile(r"\b\d+(?:\.\d+)?\s*(?:[KMG]i?B)\s*/\s*\d+(?:\.\d+)?\s*(?:[KMG]i?B)"),  # 4.2MB/10MB
    re.compile(r"\b\d+(?:\.\d+)?\s*(?:[KMG]i?B|B)/s\b"),         # 3.4MB/s
    re.compile(r"^[\s\|\-\\/\*\.=#>â–ˆâ–‘â–’â–“â–ºâ ‹â ™â ¹â ¸â ¼â ´â ¦â §â ‡â ]+$"),           # spinner / bar frames
    re.compile(r"^\s*(?:\[=*>?\s*\]|\[#*\s*\]|\[\.*\s*\])\s*$"),  # [====>   ]
    re.compile(r"^\s*(?:Progress|Status):", re.IGNORECASE),
    re.compile(r"^\s*(?:Receiving|Counting|Compressing|Delta) objects:", re.IGNORECASE),
    re.compile(r"^\s*remote:\s*(?:Counting|Compressing|Enumerating|Total)\b", re.IGNORECASE),
    re.compile(r"^\s*(?:Reading (?:package|state)|Get:\d+|Hit:\d+|Ign:\d+)\b"),
    re.compile(r"^\s*(?:npm|yarn|pnpm)\s+(?:WARN\s+)?(?:idealTree|timing|sill|http fetch)\b"),
    re.compile(r"^\s*\.{3,}\s*$"),
)

_ANSI = re.compile(r"\x1b\[[0-9;?]*[a-zA-Z]|\x1b\][^\x07]*\x07")
_CR_FRAMES = re.compile(r"^.*\r(?!\n)")  # keep only the final frame of a \r-redrawn line


def _is_signal(line: str) -> bool:
    return any(p.search(line) for p in _SIGNAL_PATTERNS)


def _is_noise(line: str) -> bool:
    return any(p.search(line) for p in _NOISE_PATTERNS)


def lossless_terminal_cleaner(
    raw_logs: str,
    *,
    annotate_drops: bool = True,
    collapse_duplicates: bool = True,
) -> str:
    """Strip build tickers and download loops, keep failures and completions.

    Signal beats noise: a line matching an error/completion pattern is kept even
    if it also looks like a progress ticker (e.g. ``[42/250] error: ...``).

    Parameters
    ----------
    annotate_drops:
        Insert a compact ``... <N> progress lines omitted ...`` marker for each
        removed run, so the model still knows work happened (and how much)
        without paying for 250 near-identical lines.
    collapse_duplicates:
        Fold runs of identical consecutive kept lines into ``line  (xN)``.
    """
    if not raw_logs:
        return ""

    out: List[str] = []
    dropped_run = 0
    prev_line: Optional[str] = None
    prev_count = 0

    def flush_duplicates() -> None:
        nonlocal prev_line, prev_count
        if prev_line is None:
            return
        if prev_count > 1:
            out.append(f"{prev_line}  (x{prev_count})")
        else:
            out.append(prev_line)
        prev_line, prev_count = None, 0

    def flush_dropped() -> None:
        nonlocal dropped_run
        if dropped_run and annotate_drops:
            out.append(f"... {dropped_run} progress lines omitted ...")
        dropped_run = 0

    for raw_line in raw_logs.splitlines():
        # Collapse carriage-return redraws down to the last rendered frame,
        # then remove ANSI colour/cursor escapes.
        line = _CR_FRAMES.sub("", raw_line)
        line = _ANSI.sub("", line).rstrip()

        if line.strip() == "":
            continue

        if _is_signal(line) or not _is_noise(line):
            flush_dropped()
            if collapse_duplicates and line == prev_line:
                prev_count += 1
            else:
                flush_duplicates()
                prev_line, prev_count = line, 1
            continue

        # Noise.
        flush_duplicates()
        dropped_run += 1

    flush_duplicates()
    flush_dropped()

    return "\n".join(out)

File: rust/src/compressors.rs
//! Lossless compressors: source comments and terminal progress noise.

use once_cell::sync::Lazy;
use regex::Regex;

/// Tuning knobs for [`lossless_code_compressor`].
#[derive(Debug, Clone)]
pub struct CodeOptions {
    pub max_consecutive_blank_lines: usize,
    pub strip_trailing_comments: bool,
    pub keep_semantic_comments: bool,
}

impl Default for CodeOptions {
    fn default() -> Self {
        Self {
            max_consecutive_blank_lines: 1,
            strip_trailing_comments: true,
            keep_semantic_comments: true,
        }
    }
}

static SEMANTIC_COMMENT: Lazy<Regex> = Lazy::new(|| {
    Regex::new(
        r"(?ix)^\#\s*(?:
              !
            | -\*-
            | (?:type|mypy|pyright)\b
            | noqa\b
            | pragma\b
            | pylint\b
            | flake8\b
            | ruff\b
            | fmt\s*:\s*(?:on|off|skip)
            | yapf\s*:\s*(?:disable|enable)
            | isort\s*:\s*
            | nosec\b
            | coding[:=]
        )",
    )
    .expect("valid semantic-comment regex")
});

/// Scan one line, tracking string state. Returns the byte index of a real
/// comment start (if any) and the triple-quote delimiter still open at EOL.
fn scan_line(line: &str, in_triple: Option<&str>) -> (Option<usize>, Option<&'static str>) {
    let bytes: Vec<char> = line.chars().collect();
    let n = bytes.len();
    let mut i = 0usize;
    // Map char index -> byte index so we can return a byte offset for slicing.
    let byte_at = |ci: usize| -> usize { line.char_indices().nth(ci).map(|(b, _)| b).unwrap_or(line.len()) };

    let mut open: Option<&'static str> = match in_triple {
        Some("\"\"\"") => Some("\"\"\""),
        Some("'''") => Some("'''"),
        _ => None,
    };

    if let Some(delim) = open {
        let d: Vec<char> = delim.chars().collect();
        while i < n {
            if bytes[i] == '\\' {
                i += 2;
                continue;
            }
            if i + d.len() <= n && bytes[i..i + d.len()] == d[..] {
                i += d.len();
                open = None;
                break;
            }
            i += 1;
        }
        if open.is_some() {
            return (None, open);
        }
    }

    while i < n {
        let ch = bytes[i];

        if ch == '#' {
            return (Some(byte_at(i)), None);
        }

        if ch == '\'' || ch == '"' {
            let triple: &'static str = if ch == '"' { "\"\"\"" } else { "'''" };
            let t: Vec<char> = triple.chars().collect();
            if i + 3 <= n && bytes[i..i + 3] == t[..] {
                i += 3;
                let mut closed = false;
                while i < n {
                    if bytes[i] == '\\' {
                        i += 2;
                        continue;
                    }
                    if i + 3 <= n && bytes[i..i + 3] == t[..] {
                        i += 3;
                        closed = true;
                        break;
                    }
                    i += 1;
                }
                if !closed {
                    return (None, Some(triple));
                }
                continue;
            }

            let quote = ch;
            i += 1;
            while i < n {
                if bytes[i] == '\\' {
                    i += 2;
                    continue;
                }
                if bytes[i] == quote {
                    i += 1;
                    break;
                }
                i += 1;
            }
            continue;
        }

        i += 1;
    }

    (None, None)
}

/// Remove human comments and collapse blank-line runs without touching
/// indentation, string contents or any executable token.
pub fn lossless_code_compressor(raw_code: &str, opts: &CodeOptions) -> String {
    if raw_code.is_empty() {
        return String::new();
    }

    let mut out: Vec<String> = Vec::new();
    let mut in_triple: Option<&'static str> = None;
    let mut blank_run = 0usize;

    for (idx, line) in raw_code.lines().enumerate() {
        let was_in_string = in_triple.is_some();
        let (comment_start, next_triple) = scan_line(line, in_triple);
        in_triple = next_triple;

        if was_in_string {
            out.push(line.to_string());
            blank_run = 0;
            continue;
        }

        let mut current: String;

        if let Some(cs) = comment_start {
            let comment_text = &line[cs..];
            let before = line[..cs].trim();
            let whole_line = before.is_empty();

            let keep = opts.keep_semantic_comments
                && (SEMANTIC_COMMENT.is_match(comment_text.trim())
                    || (idx == 0 && comment_text.starts_with("#!")));

            if keep {
                out.push(line.trim_end().to_string());
                blank_run = 0;
                continue;
            }
            if whole_line {
                continue;
            }
            current = if opts.strip_trailing_comments {
                line[..cs].trim_end().to_string()
            } else {
                line.trim_end().to_string()
            };
        } else {
            current = line.trim_end().to_string();
        }

        if current.trim().is_empty() {
            blank_run += 1;
            if blank_run > opts.max_consecutive_blank_lines {
                continue;
            }
            current = String::new();
            out.push(current);
            continue;
        }

        blank_run = 0;
        out.push(current);
    }

    while out.first().map_or(false, |l| l.trim().is_empty()) {
        out.remove(0);
    }
    while out.last().map_or(false, |l| l.trim().is_empty()) {
        out.pop();
    }

    out.join("\n")
}

// ---------------------------------------------------------------------------
// Terminal cleaning
// ---------------------------------------------------------------------------

static SIGNAL: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"(?i)\b(?:error|errors)\b",
        r"(?i)\b(?:fatal|panic|abort(?:ed)?|segfault|core dumped)\b",
        r"(?i)\b(?:fail|failed|failure|failing)\b",
        r"(?i)\b(?:exception|traceback|stack ?trace|assertion)\b",
        r"(?i)\b(?:undefined reference|cannot find|not found|no such file)\b",
        r"(?i)\b(?:exit(?:ed)? (?:code|status)|exit code)\b",
        r"(?i)\b(?:succeeded|success|successful|completed successfully)\b",
        r"(?i)\b(?:build (?:finished|complete|succeeded|failed))\b",
        r"(?i)\b(?:finished|done) in\b",
        r"^\s*(?:E|ERROR|FAIL|FAILED|CRITICAL)\b",
        r"(?i)\bwarning\b.*\b(?:treated as error|-Werror)\b",
        r#"^\s*(?:at |File "|\s+\^+\s*$)"#,
    ]
    .iter()
    .map(|p| Regex::new(p).expect("valid signal regex"))
    .collect()
});

static NOISE: Lazy<Vec<Regex>> = Lazy::new(|| {
    [
        r"^\s*\[\s*\d+\s*/\s*\d+\s*\]",
        r"^\s*\(\s*\d+\s*/\s*\d+\s*\)",
        r"^\s*\d{1,3}\s*%",
        r"(?i)\b\d{1,3}(?:\.\d+)?\s*%\s*(?:complete|completed|done|finished)?\b",
        r"(?i)^\s*(?:Compiling|Building|Downloading|Fetching|Extracting|Unpacking|Installing|Resolving|Linking|Indexing|Uploading|Pulling|Cloning)\b",
        r"\b(?:ETA|eta)\s+\d",
        r"\b\d+(?:\.\d+)?\s*(?:[KMG]i?B)\s*/\s*\d+(?:\.\d+)?\s*(?:[KMG]i?B)",
        r"\b\d+(?:\.\d+)?\s*(?:[KMG]i?B|B)/s\b",
        r"^[\s\|\-\\/\*\.=#>█░▒▓►⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏]+$",
        r"^\s*(?:\[=*>?\s*\]|\[#*\s*\]|\[\.*\s*\])\s*$",
        r"(?i)^\s*(?:Progress|Status):",
        r"(?i)^\s*(?:Receiving|Counting|Compressing|Delta) objects:",
        r"(?i)^\s*remote:\s*(?:Counting|Compressing|Enumerating|Total)\b",
        r"^\s*(?:Reading (?:package|state)|Get:\d+|Hit:\d+|Ign:\d+)\b",
        r"^\s*(?:npm|yarn|pnpm)\s+(?:WARN\s+)?(?:idealTree|timing|sill|http fetch)\b",
        r"^\s*\.{3,}\s*$",
    ]
    .iter()
    .map(|p| Regex::new(p).expect("valid noise regex"))
    .collect()
});

static ANSI: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"\x1b\[[0-9;?]*[a-zA-Z]|\x1b\][^\x07]*\x07").expect("valid ansi regex"));

fn is_signal(line: &str) -> bool {
    SIGNAL.iter().any(|r| r.is_match(line))
}

fn is_noise(line: &str) -> bool {
    NOISE.iter().any(|r| r.is_match(line))
}

/// Strip build tickers and download loops; keep failures and completions.
pub fn lossless_terminal_cleaner(raw_logs: &str, annotate_drops: bool, collapse_duplicates: bool) -> String {
    if raw_logs.is_empty() {
        return String::new();
    }

    let mut out: Vec<String> = Vec::new();
    let mut dropped = 0usize;
    let mut prev: Option<String> = None;
    let mut prev_count = 0usize;

    macro_rules! flush_dupes {
        () => {
            if let Some(p) = prev.take() {
                if prev_count > 1 {
                    out.push(format!("{}  (x{})", p, prev_count));
                } else {
                    out.push(p);
                }
                prev_count = 0;
            }
        };
    }

    for raw in raw_logs.lines() {
        // Keep only the final frame of a \r-redrawn line.
        let last_frame = raw.rsplit('\r').next().unwrap_or(raw);
        let cleaned = ANSI.replace_all(last_frame, "");
        let line = cleaned.trim_end();

        if line.trim().is_empty() {
            continue;
        }

        if is_signal(line) || !is_noise(line) {
            if dropped > 0 && annotate_drops {
                flush_dupes!();
                out.push(format!("... {} progress lines omitted ...", dropped));
            }
            dropped = 0;

            if collapse_duplicates && prev.as_deref() == Some(line) {
                prev_count += 1;
            } else {
                flush_dupes!();
                prev = Some(line.to_string());
                prev_count = 1;
            }
            continue;
        }

        flush_dupes!();
        dropped += 1;
    }

    flush_dupes!();
    if dropped > 0 && annotate_drops {
        out.push(format!("... {} progress lines omitted ...", dropped));
    }

    out.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hash_inside_string_survives() {
        let src = "note = \"refund # not a comment\"  # real comment\n";
        let out = lossless_code_compressor(src, &CodeOptions::default());
        assert_eq!(out, "note = \"refund # not a comment\"");
    }

    #[test]
    fn whole_line_comments_removed() {
        let src = "# chatter\nx = 1\n";
        assert_eq!(lossless_code_compressor(src, &CodeOptions::default()), "x = 1");
    }

    #[test]
    fn semantic_comments_kept() {
        let src = "#!/usr/bin/env python3\n# chatter\nimport os  # noqa\n";
        let out = lossless_code_compressor(src, &CodeOptions::default());
        assert!(out.contains("#!/usr/bin/env python3"));
        assert!(out.contains("noqa"));
        assert!(!out.contains("chatter"));
    }

    #[test]
    fn indentation_preserved() {
        let src = "if x:\n    if y:\n        deep()  # go\n";
        let out = lossless_code_compressor(src, &CodeOptions::default());
        assert!(out.contains("        deep()"));
    }

    #[test]
    fn blank_runs_collapse() {
        let src = "a = 1\n\n\n\n\nb = 2\n";
        assert_eq!(lossless_code_compressor(src, &CodeOptions::default()), "a = 1\n\nb = 2");
    }

    #[test]
    fn triple_quoted_block_untouched() {
        let src = "x = \"\"\"a\n\n\n\nb\"\"\"\n";
        let out = lossless_code_compressor(src, &CodeOptions::default());
        assert!(out.contains("a\n\n\n\nb"));
    }

    #[test]
    fn errors_survive_cleaning() {
        let logs = "[1/250] Compiling a.c\n 45% completed\nsrc/pool.c:88: error: bad member\nBuild failed in 48.2s\n";
        let out = lossless_terminal_cleaner(logs, true, true);
        assert!(out.contains("error: bad member"));
        assert!(out.contains("Build failed"));
        assert!(!out.contains("Compiling a.c"));
    }

    #[test]
    fn signal_beats_noise() {
        let logs = "[42/250] error: on fire\n[43/250] Compiling ok.c\n";
        let out = lossless_terminal_cleaner(logs, true, true);
        assert!(out.contains("on fire"));
        assert!(!out.contains("ok.c"));
    }

    #[test]
    fn ansi_stripped() {
        let out = lossless_terminal_cleaner("\x1b[31mfatal: broken\x1b[0m\n", true, true);
        assert_eq!(out, "fatal: broken");
    }

    #[test]
    fn duplicates_collapse() {
        let out = lossless_terminal_cleaner("same error\nsame error\nsame error\n", true, true);
        assert_eq!(out, "same error  (x3)");
    }

    #[test]
    fn empty_inputs() {
        assert_eq!(lossless_code_compressor("", &CodeOptions::default()), "");
        assert_eq!(lossless_terminal_cleaner("", true, true), "");
    }
}
